### PR TITLE
Rectify: Guard Skill-Exemption Structural Enforcement

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

The `pr_create_guard` blocks legitimate pipeline skills (compose-pr, compose-research-pr, open-integration-pr, promote-to-main, pipeline-summary) because `HookDef` has no declarative vocabulary for skill-name exemptions. The guard's tests pass in isolation because they never inject `AUTOSKILLIT_SKILL_NAME` — the pipeline-caller scenario is architecturally invisible to the test suite. The fix adds a structural enforcement layer (analogous to `session_scope` enforcement) that makes it impossible to register a guard with `exempt_skills` without (a) implementing the env-var check in the guard source and (b) having test coverage for both blocked and exempt paths.

Key implementation steps:
1. Add `exempt_skills` field to `HookDef` (frozen dataclass, frozenset[str], default frozenset())
2. Split combined HookDef entry for `unsafe_install_guard` and `pr_create_guard` into separate entries
3. Add `AUTOSKILLIT_SKILL_NAME` check to `pr_create_guard.py` against `_EXEMPT_SKILLS` frozenset
4. Add `exempt_skills` to `_canonical_registry_payload` output
5. Create `tests/infra/test_skill_exemption_enforcement.py` (structural enforcement tests)
6. Update `tests/infra/test_pr_create_guard.py` with `TestExemptSkills` class

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/remediation-20260516-215802-020485/.autoskillit/temp/rectify/rectify_guard_exempt_skills_enforcement_2026-05-16_215802.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | Model | count | uncached | output | cache_read | peak_ctx | turns | cache_write | time |
|------|-------|-------|----------|--------|------------|----------|-------|-------------|------|
| rectify | claude-sonnet-4-6 | 1 | 72 | 15.3k | 0 | 101.0k | 245 | 0 | 14m 55s |
| dry_walkthrough | claude-opus-4-6 | 2 | 78 | 19.0k | 0 | 72.4k | 149 | 0 | 11m 34s |
| implement* | MiniMax-M2.7-highspeed | 3 | 27.9M | 151.4k | 0 | 89.5k | 855 | 0 | 1h 20m |
| prepare_pr* | MiniMax-M2.7-highspeed | 1 | 101.9k | 5.2k | 0 | 28.9k | 22 | 0 | 1m 41s |
| compose_pr* | MiniMax-M2.7-highspeed | 1 | 83.3k | 1.7k | 0 | 28.9k | 22 | 0 | 52s |
| review_pr | claude-sonnet-4-6 | 3 | 346 | 78.0k | 0 | 107.7k | 197 | 0 | 24m 27s |
| resolve_review | claude-sonnet-4-6 | 3 | 3.1k | 96.8k | 0 | 136.6k | 444 | 0 | 46m 27s |
| **Total** | | | 28.0M | 367.4k | 0 | 136.6k | | 0 | 3h 0m |

\* *Step used a non-Anthropic provider; caching behavior may differ.*

## Token Efficiency

| Step | LoC Changed | cache_read/LoC | cache_write/LoC | output/LoC |
|------|-------------|----------------|-----------------|------------|
| rectify | 0 | — | — | — |
| dry_walkthrough | 0 | — | — | — |
| implement | 499 | 0.0 | 0.0 | 303.5 |
| prepare_pr | 0 | — | — | — |
| compose_pr | 0 | — | — | — |
| review_pr | 0 | — | — | — |
| resolve_review | 214 | 0.0 | 0.0 | 452.6 |
| **Total** | **713** | 0.0 | 0.0 | 515.3 |
## Model Usage Breakdown

| Model | steps | uncached | output | cache_read | cache_write | time |
|-------|-------|----------|--------|------------|-------------|------|
| claude-sonnet-4-6 | 2 | 398 | 31.7k | 0 | 0 | 26m 49s |
| claude-opus-4-6 | 1 | 78 | 19.0k | 0 | 0 | 11m 34s |
| MiniMax-M2.7-highspeed | 4 | 48.1M | 303.8k | 0 | 0 | 2h 25m |

### CODE REVIEW

File: src/autoskillit/hook_registry.py [hunk 1]
```diff
@@ -67,7 +68,22 @@ def __post_init__(self) -> None:
     ),
     HookDef(
         matcher=r"Bash|mcp__.*autoskillit.*__run_cmd",
-        scripts=["guards/unsafe_install_guard.py", "guards/pr_create_guard.py"],
+        scripts=["guards/unsafe_install_guard.py"],
+    ),
+    HookDef(
+        matcher=r"Bash|mcp__.*autoskillit.*__run_cmd",
+        scripts=["guards/pr_create_guard.py"],
+        # Must stay in sync with _EXEMPT_SKILLS in guards/pr_create_guard.py —
```
Comment (Trecek) [inline src/autoskillit/hook_registry.py:76]: [warning] arch: Two parallel frozensets (_EXEMPT_SKILLS in pr_create_guard.py and exempt_skills on the HookDef) must stay in sync via comment only. The stdlib-only boundary prevents a shared import, so any future edit to one set without updating the other silently diverges the guard from its registration metadata.

File: src/autoskillit/recipe/rules/rules_optional_capture.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+
+def _has_optional_capture_group(patterns: list[str]) -> bool:
+    """Return True if any pattern contains a fully-optional capture group.
+
+    A fully-optional capture group is a (...) expression followed by ?
+    at the end of the pattern, e.g. ``(https://...)?`` makes the entire URL optional.
+    """
+    for pattern in patterns:
+        if re.search(r"\([^)]+\)\?$", pattern):
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_optional_capture.py]: [warning] arch: The regex `r"\([^)]+\)\?$"` anchors to end-of-string with `$`, so it only detects an optional group at the very end of a pattern string. Patterns like `(foo)?bar` or patterns with trailing anchors (e.g., `(https://...)?$`) will not be detected. If the intent is to catch any optional group anywhere in a pattern, remove the `$` anchor.

File: src/autoskillit/recipe/rules/rules_optional_capture.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        current = to_visit.pop(0)
+        if current in visited:
+            continue
+        visited.add(current)
+        step = steps.get(current)
+        if step is None:
+            continue
+
+        # Check if this step is a guard: action=route with on_result conditions
+        if step.action == "route" and step.on_result and step.on_result.conditions:
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_optional_capture.py:62]: [warning] bugs: The route-step guard detection at this branch checks `cond.when` for `context.{captured_key}`, but a downstream route step might guard on `result.{captured_key}` instead. The parallel branch at line 71 only checks `result.*` for non-route steps. A route step whose condition is `result.{captured_key}` falls through both branches undetected, producing a false-positive violation.

File: src/autoskillit/recipe/rules/rules_optional_capture.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        # The step must route somewhere — either on_success or on_result
+        routes_via_on_result = bool(step.on_result and step.on_result.conditions)
+        if not step.on_success and not routes_via_on_result:
+            continue
+
+        route_target = step.on_success or "(via on_result)"
+
+        # For each captured key, check whether a guard is interposed
+        for captured_key in step.capture:
+            if not _has_guard_for_key(step_name, captured_key, ctx.recipe.steps):
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_optional_capture.py:136]: [warning] bugs: _has_guard_for_key is called with step_name = the producer step itself. The BFS starts by visiting the producer: it is not a route step (L57 branch is skipped), so it falls into the on_result/on_success branch and enqueues downstream steps. This is correct for the current usage, but the function silently treats the producer step as a potential guard location on first visit. Add a docstring clarification or start BFS from the producer's immediate successors to make the intent explicit.

File: tests/infra/test_skill_exemption_enforcement.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    pairs = []
+    for hookdef in HOOK_REGISTRY:
+        exempt = getattr(hookdef, "exempt_skills", frozenset())
+        if exempt:
+            for script in hookdef.scripts:
+                pairs.append((hookdef, script))
+    return pairs
+
+
+@pytest.mark.parametrize("hookdef,script", _exempt_hooks())
```
Comment (Trecek) [inline tests/infra/test_skill_exemption_enforcement.py:62]: [warning] tests: `@pytest.mark.parametrize('hookdef,script', _exempt_hooks())` collects zero test cases if no HookDef in HOOK_REGISTRY has `exempt_skills` set, causing `test_exempt_guard_contains_skill_name_check` to vacuously pass. Add a separate test asserting `len(_exempt_hooks()) >= 1` to ensure the parametrize never runs with an empty list.

File: tests/infra/test_skill_exemption_enforcement.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    stem = Path(guard_script).stem  # e.g. "pr_create_guard"
+    tests_infra = Path(__file__).resolve().parent
+    for directory in (tests_infra, tests_infra.parent / "hooks"):
+        candidate = directory / f"test_{stem}.py"
+        if candidate.exists():
+            return candidate
+    return None
+
+
+@pytest.mark.parametrize("guard_script", _exempt_guard_names())
```
Comment (Trecek) [inline tests/infra/test_skill_exemption_enforcement.py:108]: [warning] tests: `@pytest.mark.parametrize('guard_script', _exempt_guard_names())` has the same vacuous-pass risk: if `_exempt_guard_names()` returns an empty list, `test_exempt_guard_has_exempt_path_test_cases` collects zero cases and silently passes. Add a guard assertion that at least one exempt guard name exists.

File: src/autoskillit/recipe/skill_contracts.yaml [hunk 7]
```diff
@@ -50,7 +50,7 @@ skills:
     - name: is_fixable
       type: string
     expected_output_patterns:
-    - "diagnosis_path\\s*=\\s*(?:/.+)?"
+    - "diagnosis_path\\s*="
```
Comment (Trecek) [inline src/autoskillit/recipe/skill_contracts.yaml]: [warning] bugs: The replacement pattern `"diagnosis_path\\s*="` is less selective than the removed `"diagnosis_path\\s*=\\s*(?:/.+)?"`. The old pattern required the value (if non-empty) to start with `/`, restricting to absolute paths or empty. The new pattern matches any value after `=` including relative paths or malformed values. If the contract verifies that diagnosis_path is either an absolute path or empty, the value constraint should be preserved.

File: src/autoskillit/recipe/__init__.py [hunk 8]
```diff
@@ -116,6 +116,9 @@
 from autoskillit.recipe.rules import rules_loop_progress as _rules_loop_progress  # noqa: E402 F401
 from autoskillit.recipe.rules import rules_merge as _rules_merge  # noqa: E402 F401
 from autoskillit.recipe.rules import rules_merge_queue as _rules_merge_queue  # noqa: E402 F401
+from autoskillit.recipe.rules import (
```
Comment (Trecek) [inline src/autoskillit/recipe/__init__.py:119]: [warning] cohesion: Multi-line parenthesized import form for rules_optional_capture breaks the single-line pattern used by all adjacent rules imports (rules_loop_progress, rules_merge, rules_merge_queue, etc.). Use: `from autoskillit.recipe.rules import rules_optional_capture as _rules_optional_capture  # noqa: E402 F401`

File: tests/recipe/test_remediation_pr_decomposition.py [hunk 9]
```diff
@@ -40,9 +40,21 @@ def test_run_arch_lenses_routes_to_compose_on_failure(recipe):
     assert step.on_failure == "compose_pr"
 
 
-def test_compose_pr_routes_to_extract_pr_number(recipe):
+def test_compose_pr_routes_to_guard_pr_url(recipe):
```
Comment (Trecek) [inline tests/recipe/test_remediation_pr_decomposition.py:43]: [warning] cohesion: test_compose_pr_routes_to_guard_pr_url only verifies conditions[0] (truthy route to guard_pr_url) but does not assert the else-branch routes to release_issue_failure, unlike the structurally equivalent test in test_bundled_recipes_pipeline_structure.py which checks both conditions. The missing else-branch assertion could hide routing regressions.

File: src/autoskillit/hooks/guards/pr_create_guard.py [hunk 10]
```diff
@@ -25,6 +26,18 @@
     "Direct gh pr create bypasses mandatory arch-lens, annotation, and review steps."
 )
 
+# Must stay in sync with the exempt_skills frozenset on the pr_create_guard HookDef
+# in hook_registry.py — stdlib-only boundary prevents a shared import.
+_EXEMPT_SKILLS: frozenset[str] = frozenset(
```
Comment (Trecek) [inline src/autoskillit/hooks/guards/pr_create_guard.py:31]: [warning] cohesion: _EXEMPT_SKILLS is duplicated from the exempt_skills frozenset on the HookDef in hook_registry.py with only a sync comment for enforcement. The stdlib-only boundary prevents a shared import, but a CI-visible mechanism (e.g., a test in test_skill_exemption_enforcement.py comparing the two sets) would catch drift before it ships.

File: src/autoskillit/recipe/rules/rules_contracts.py [hunk 11]
```diff
@@ -253,10 +253,12 @@ def _check_write_behavior_consistency(ctx: ValidationContext) -> list[RuleFindin
         "no changes needed",
         "already green",
         "graceful degradation",  # resolve-review: "graceful degradation — do not fail"
+        r"degrade\s+gracefully",  # compose-pr: "Degrade gracefully when ..."
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_contracts.py]: [warning] slop: Inline comment `# compose-pr: "Degrade gracefully when ..."` references the calling skill by name. Per project conventions, comments must not reference callers — this belongs in the PR description. The comment will go stale if compose-pr's wording changes.

File: src/autoskillit/recipe/rules/rules_contracts.py [hunk 12]
```diff
@@ -253,10 +253,12 @@ def _check_write_behavior_consistency(ctx: ValidationContext) -> list[RuleFindin
...
         r"\bskip\b.{0,30}\bstep\b",  # conditional path language
         r"(?:and|,)\s*exit\s+0",  # graceful early exit suffix (e.g. "and exit 0", ", exit 0")
         r"\bif no\b.{0,30}\bfound\b",
-        r"if.*unavailable",
+        r"if.*unavailable",  # compose-pr: "if GitHub unavailable"
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_contracts.py]: [warning] slop: Inline comment `# compose-pr: "if GitHub unavailable"` references the calling skill by name. Per project conventions, caller references belong in the PR description, not inline comments.

File: src/autoskillit/recipe/rules/rules_contracts.py [hunk 13]
```diff
@@ -253,10 +253,12 @@ def _check_write_behavior_consistency(ctx: ValidationContext) -> list[RuleFindin
...
+        r"when.*unavailable",  # compose-pr: "when GitHub unavailable"
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_contracts.py]: [warning] slop: Inline comment `# compose-pr: "when GitHub unavailable"` references the calling skill by name. Per project conventions, caller references belong in the PR description, not inline comments.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: approved_with_comments

File: src/autoskillit/hook_registry.py (hunk 1 above)
Comment (Trecek) [inline src/autoskillit/hook_registry.py:76]: [warning] arch: Two parallel frozensets (_EXEMPT_SKILLS in pr_create_guard.py and exempt_skills on the HookDef) must stay in sync via comment only. The stdlib-only boundary prevents a shared import, so any future edit to one set without updating the other silently diverges the guard from its registration metadata.

File: src/autoskillit/recipe/rules/rules_optional_capture.py (hunk 2 above)
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_optional_capture.py]: [warning] arch: The regex `r"\([^)]+\)\?$"` anchors to end-of-string with `$`, so it only detects an optional group at the very end of a pattern string. Patterns like `(foo)?bar` or patterns with trailing anchors (e.g., `(https://...)?$`) will not be detected. If the intent is to catch any optional group anywhere in a pattern, remove the `$` anchor.

File: src/autoskillit/recipe/rules/rules_optional_capture.py (hunk 3 above)
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_optional_capture.py:62]: [warning] bugs: The route-step guard detection at this branch checks `cond.when` for `context.{captured_key}`, but a downstream route step might guard on `result.{captured_key}` instead. The parallel branch at line 71 only checks `result.*` for non-route steps. A route step whose condition is `result.{captured_key}` falls through both branches undetected, producing a false-positive violation.

File: src/autoskillit/recipe/rules/rules_optional_capture.py (hunk 4 above)
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_optional_capture.py:136]: [warning] bugs: _has_guard_for_key is called with step_name = the producer step itself. The BFS starts by visiting the producer: it is not a route step (L57 branch is skipped), so it falls into the on_result/on_success branch and enqueues downstream steps. This is correct for the current usage, but the function silently treats the producer step as a potential guard location on first visit. Add a docstring clarification or start BFS from the producer's immediate successors to make the intent explicit.

File: tests/infra/test_skill_exemption_enforcement.py (hunk 5 above)
Comment (Trecek) [inline tests/infra/test_skill_exemption_enforcement.py:62]: [warning] tests: `@pytest.mark.parametrize('hookdef,script', _exempt_hooks())` collects zero test cases if no HookDef in HOOK_REGISTRY has `exempt_skills` set, causing `test_exempt_guard_contains_skill_name_check` to vacuously pass. Add a separate test asserting `len(_exempt_hooks()) >= 1` to ensure the parametrize never runs with an empty list.

File: tests/infra/test_skill_exemption_enforcement.py (hunk 6 above)
Comment (Trecek) [inline tests/infra/test_skill_exemption_enforcement.py:108]: [warning] tests: `@pytest.mark.parametrize('guard_script', _exempt_guard_names())` has the same vacuous-pass risk: if `_exempt_guard_names()` returns an empty list, `test_exempt_guard_has_exempt_path_test_cases` collects zero cases and silently passes. Add a guard assertion that at least one exempt guard name exists.

File: src/autoskillit/recipe/skill_contracts.yaml (hunk 7 above)
Comment (Trecek) [inline src/autoskillit/recipe/skill_contracts.yaml]: [warning] bugs: The replacement pattern `"diagnosis_path\\s*="` is less selective than the removed `"diagnosis_path\\s*=\\s*(?:/.+)?"`. The old pattern required the value (if non-empty) to start with `/`, restricting to absolute paths or empty. The new pattern matches any value after `=` including relative paths or malformed values. If the contract verifies that diagnosis_path is either an absolute path or empty, the value constraint should be preserved.

File: src/autoskillit/recipe/__init__.py (hunk 8 above)
Comment (Trecek) [inline src/autoskillit/recipe/__init__.py:119]: [warning] cohesion: Multi-line parenthesized import form for rules_optional_capture breaks the single-line pattern used by all adjacent rules imports (rules_loop_progress, rules_merge, rules_merge_queue, etc.). Use: `from autoskillit.recipe.rules import rules_optional_capture as _rules_optional_capture  # noqa: E402 F401`

File: tests/recipe/test_remediation_pr_decomposition.py (hunk 9 above)
Comment (Trecek) [inline tests/recipe/test_remediation_pr_decomposition.py:43]: [warning] cohesion: test_compose_pr_routes_to_guard_pr_url only verifies conditions[0] (truthy route to guard_pr_url) but does not assert the else-branch routes to release_issue_failure, unlike the structurally equivalent test in test_bundled_recipes_pipeline_structure.py which checks both conditions. The missing else-branch assertion could hide routing regressions.

File: src/autoskillit/hooks/guards/pr_create_guard.py (hunk 10 above)
Comment (Trecek) [inline src/autoskillit/hooks/guards/pr_create_guard.py:31]: [warning] cohesion: _EXEMPT_SKILLS is duplicated from the exempt_skills frozenset on the HookDef in hook_registry.py with only a sync comment for enforcement. The stdlib-only boundary prevents a shared import, but a CI-visible mechanism (e.g., a test in test_skill_exemption_enforcement.py comparing the two sets) would catch drift before it ships.

File: src/autoskillit/recipe/rules/rules_contracts.py (hunk 11 above)
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_contracts.py]: [warning] slop: Inline comment `# compose-pr: "Degrade gracefully when ..."` references the calling skill by name. Per project conventions, comments must not reference callers — this belongs in the PR description. The comment will go stale if compose-pr's wording changes.

File: src/autoskillit/recipe/rules/rules_contracts.py (hunk 12 above)
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_contracts.py]: [warning] slop: Inline comment `# compose-pr: "if GitHub unavailable"` references the calling skill by name. Per project conventions, caller references belong in the PR description, not inline comments.

File: src/autoskillit/recipe/rules/rules_contracts.py (hunk 13 above)
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_contracts.py]: [warning] slop: Inline comment `# compose-pr: "when GitHub unavailable"` references the calling skill by name. Per project conventions, caller references belong in the PR description, not inline comments.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: approved_with_comments

**Review by Trecek — COMMENTED** [review]
AutoSkillit review: warning-only findings detected. See inline comments — no blocking changes required.

File: tests/recipe/test_optional_capture_guard_rule.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    monkeypatch.setattr(_r, "load_bundled_manifest", lambda: manifest)
+
+    recipe = _make_recipe(
+        {
+            "producer": RecipeStep(
+                tool="run_skill",
+                with_args={"skill_command": "/autoskillit:test-skill"},
+                capture={"pr_url": "${{ result.pr_url }}"},
+                on_success="consumer",
+            ),
```
Comment (Trecek) [inline tests/recipe/test_optional_capture_guard_rule.py:146]: **Observation from local review round 1:**

test_implementation_yaml_no_diagnostic_after_guard_added is marked pytest.mark.small but performs real filesystem access via pkg_root() and full recipe + semantic-rule evaluation. This is an integration-level test and should carry pytest.mark.medium (or at minimum a comment explaining the small classification).

**Evidence:** tests/CLAUDE.md defines small as 'No persistent I/O, no network, no subprocess.' But the established codebase pattern marks bundled-YAML read tests as small — conflicts with CLAUDE.md definition. Human decision needed on whether to enforce strictly or document the exception.

**Review by Trecek — COMMENTED** [review]
Observations accumulated from 1 local review rounds:

File: src/autoskillit/hook_registry.py (hunk 1 above)
Comment (Trecek) [reply src/autoskillit/hook_registry.py:76]: Investigated — this is intentional. The stdlib-only boundary makes a shared import impossible; sync comments document the constraint. A CI test comparing both frozensets was added in 05740314 — drift is now caught automatically.

Comment (Trecek) [reply src/autoskillit/hook_registry.py:76]: Investigated — this is intentional. The stdlib-only boundary makes a shared import impossible; sync comments document the constraint. A CI test comparing both frozensets was added in 05740314 — drift is now caught automatically.

File: src/autoskillit/recipe/rules/rules_optional_capture.py (hunk 2 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_optional_capture.py]: Investigated — this is intentional. The docstring explicitly documents the 'end of pattern' constraint: 'a (...) expression followed by ? at the end of the pattern'. The $ anchor is intentional design. Separately, non-capturing group (?:...) false-positives were fixed in 3693070f by excluding (?:...) from the detection regex.

Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_optional_capture.py]: Investigated — this is intentional. The docstring explicitly documents the 'end of pattern' constraint: 'a (...) expression followed by ? at the end of the pattern'. The $ anchor is intentional design. Separately, non-capturing group (?:...) false-positives were fixed in 3693070f by excluding (?:...) from the detection regex.

File: src/autoskillit/recipe/rules/rules_optional_capture.py (hunk 3 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_optional_capture.py:62]: Agreed — fixed in 024332d1. Route step guard detection now also checks result.{captured_key} in addition to context.{captured_key}.

Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_optional_capture.py:62]: Agreed — fixed in 024332d1. Route step guard detection now also checks result.{captured_key} in addition to context.{captured_key}.

File: src/autoskillit/recipe/rules/rules_optional_capture.py (hunk 4 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_optional_capture.py:136]: Agreed — fixed in 024332d1. Docstring now explicitly documents that BFS starts at the producer step itself and explains why this is correct.

Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_optional_capture.py:136]: Agreed — fixed in 024332d1. Docstring now explicitly documents that BFS starts at the producer step itself and explains why this is correct.

File: tests/infra/test_skill_exemption_enforcement.py (hunk 5 above)
Comment (Trecek) [reply tests/infra/test_skill_exemption_enforcement.py:62]: Agreed — fixed in 05740314. Added test_at_least_one_hookdef_has_exempt_skills() asserting len(_exempt_hooks()) >= 1.

Comment (Trecek) [reply tests/infra/test_skill_exemption_enforcement.py:62]: Agreed — fixed in 05740314. Added test_at_least_one_hookdef_has_exempt_skills() asserting len(_exempt_hooks()) >= 1.

File: tests/infra/test_skill_exemption_enforcement.py (hunk 6 above)
Comment (Trecek) [reply tests/infra/test_skill_exemption_enforcement.py:108]: Agreed — fixed in 05740314. Added test_at_least_one_exempt_guard_name_exists() asserting len(_exempt_guard_names()) >= 1.

Comment (Trecek) [reply tests/infra/test_skill_exemption_enforcement.py:108]: Agreed — fixed in 05740314. Added test_at_least_one_exempt_guard_name_exists() asserting len(_exempt_guard_names()) >= 1.

File: src/autoskillit/recipe/skill_contracts.yaml (hunk 7 above)
Comment (Trecek) [reply src/autoskillit/recipe/skill_contracts.yaml]: Agreed — fixed in 3b8d529f. Restored pattern to diagnosis_path\s*=\s*(?:/.+)? — enforces absolute-path-or-empty. Also fixed _has_optional_capture_group in 3693070f to exclude non-capturing groups from detection.

Comment (Trecek) [reply src/autoskillit/recipe/skill_contracts.yaml]: Agreed — fixed in 3b8d529f. Restored pattern to diagnosis_path\s*=\s*(?:/.+)? — enforces absolute-path-or-empty. Also fixed _has_optional_capture_group in 3693070f to exclude non-capturing groups from detection.

File: src/autoskillit/recipe/__init__.py (hunk 8 above)
Comment (Trecek) [reply src/autoskillit/recipe/__init__.py:119]: Investigated — this is intentional. Attempted — formatter enforces multi-line for this specific import: the single-line form is 106 chars vs 99-char limit (pyproject.toml line-length = 99). Ruff wraps it back. No change applied.

Comment (Trecek) [reply src/autoskillit/recipe/__init__.py:119]: Investigated — this is intentional. Attempted — formatter enforces multi-line for this specific import: the single-line form is 106 chars vs 99-char limit (pyproject.toml line-length = 99). Ruff wraps it back. No change applied.

File: tests/recipe/test_remediation_pr_decomposition.py (hunk 9 above)
Comment (Trecek) [reply tests/recipe/test_remediation_pr_decomposition.py:43]: Agreed — fixed in 918f9475. Added else-branch assertion: conditions[-1].route == 'release_issue_failure'. Also tightened >= 1 to >= 2 to enforce both conditions are present.

Comment (Trecek) [reply tests/recipe/test_remediation_pr_decomposition.py:43]: Agreed — fixed in 918f9475. Added else-branch assertion: conditions[-1].route == 'release_issue_failure'. Also tightened >= 1 to >= 2 to enforce both conditions are present.

File: src/autoskillit/hooks/guards/pr_create_guard.py (hunk 10 above)
Comment (Trecek) [reply src/autoskillit/hooks/guards/pr_create_guard.py:31]: Agreed — fixed in 05740314. Added test_pr_create_guard_exempt_skills_matches_hookdef() in test_skill_exemption_enforcement.py — loads pr_create_guard.py via importlib and compares _EXEMPT_SKILLS with HookDef.exempt_skills.

Comment (Trecek) [reply src/autoskillit/hooks/guards/pr_create_guard.py:31]: Agreed — fixed in 05740314. Added test_pr_create_guard_exempt_skills_matches_hookdef() in test_skill_exemption_enforcement.py — loads pr_create_guard.py via importlib and compares _EXEMPT_SKILLS with HookDef.exempt_skills.

File: src/autoskillit/recipe/rules/rules_contracts.py (hunk 11 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_contracts.py]: Agreed — fixed in 92b99077. Removed compose-pr caller reference from the inline comment.

Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_contracts.py]: Agreed — fixed in 92b99077. Removed compose-pr caller reference from the inline comment.

File: src/autoskillit/recipe/rules/rules_contracts.py (hunk 12 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_contracts.py]: Agreed — fixed in 92b99077. Removed compose-pr caller reference from the inline comment.

Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_contracts.py]: Agreed — fixed in 92b99077. Removed compose-pr caller reference from the inline comment.

File: src/autoskillit/recipe/rules/rules_contracts.py (hunk 13 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_contracts.py]: Agreed — fixed in 92b99077. Removed compose-pr caller reference from the inline comment.

Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_contracts.py]: Agreed — fixed in 92b99077. Removed compose-pr caller reference from the inline comment.

File: tests/recipe/test_optional_capture_guard_rule.py (hunk 14 above)
Comment (Trecek) [reply tests/recipe/test_optional_capture_guard_rule.py:146]: Valid observation — flagged for design decision. Valid observation — the established codebase pattern marks bundled-YAML read tests as small even though CLAUDE.md says no persistent I/O. Human decision needed: enforce strictly or document the exception.
